tools/envoy_collect/envoy_collect.py

```diff
@@ -0,0 +1,243 @@
+  if perf:
+    return config_path, []
+
+  # Load original Envoy config.
+  with open(config_path, 'r') as f:
+    envoy_config = json.loads(f.read())
+
+  # Add unconditional access logs for all listeners.
+  access_log_paths = []
+  for n, listener in enumerate(envoy_config['listeners']):
+    for network_filter in listener['filters']:
+      if network_filter['name'] == 'http_connection_manager':
+        config = network_filter['config']
+        access_log_path = os.path.join(output_directory, 'access_%d.log' % n)
+        access_log_config = {'path': access_log_path}
+        if 'access_log' in config:
+          config['access_log'].append(access_log_config)
+        else:
+          config['access_log'] = [access_log_config]
+        access_log_paths.append(access_log_path)
+
+  # Write out modified Envoy config.
+  modified_envoy_config_path = os.path.join(output_directory, 'config.json')
+  with open(modified_envoy_config_path, 'w') as f:
+    f.write(json.dumps(envoy_config, indent=2))
+
+  return modified_envoy_config_path, access_log_paths
+
+
+def EnvoyCollect(parse_result, unknown_args):
+  """Run Envoy and collect its artifacts.
+
+  Args:
+    parse_result: Namespace object with envoy_collect.py's args.
+    unknown_args: list of remaining args to pass to Envoy binary.
+  """
+  # Are we in performance mode? Otherwise, debug.
+  perf = parse_result.performance
+  envoy_tmpdir = tempfile.mkdtemp()
+  return_code = 1  # Non-zero default return.
+  # Try and do stuff with envoy_tmpdir, rm -rf regardless of success/failure.
+  try:
+    # Setup Envoy config and determine the paths of the files we're going to
+    # generate.
+    modified_envoy_config_path, access_log_paths = ModifyEnvoyconfig(
+        parse_result.config_path, perf, envoy_tmpdir)
+    dump_handlers_paths = {
+        h: os.path.join(envoy_tmpdir, '%s.txt' % h)
+        for h in DUMP_HANDLERS
+    }
+    envoy_log_path = os.path.join(envoy_tmpdir, 'envoy.log')
+    # The manifest of files that will be placed in the output .tar.
+    manifest = access_log_paths + list(dump_handlers_paths.values()) + [
+        modified_envoy_config_path, envoy_log_path
+    ]
+    # This is where we will find out where the admin endpoint is listening.
+    admin_address_path = os.path.join(envoy_tmpdir, 'admin_address.txt')
+
+    # Only run under 'perf record' in performance mode.
+    if perf:
+      perf_data_path = os.path.join(envoy_tmpdir, 'perf.data')
+      manifest.append(perf_data_path)
+      perf_record_args = [
+          PERF_PATH,
+          'record',
+          '-o',
+          perf_data_path,
+          '-g',
+          '--',
+      ]
+    else:
+      perf_record_args = []
+
+    # This is how we will invoke the wrapped envoy.
+    envoy_shcmd = ' '.join(
+        map(pipes.quote, perf_record_args + [
+            parse_result.envoy_binary,
+            '-c',
+            modified_envoy_config_path,
+            '-l',
+            'error' if perf else 'trace',
```

Comment:
I'll allow the user specified `-l` to take precedence if given, otherwise will default to trace in debug mode.